src/aspire/abinitio/commonline_sdp.py

```diff
@@ -1,30 +1,249 @@
+        :param A: 3 * n_img sparse arrays of constraint data.
+        :param b: 3 * n_img scalars such that tr(A_i G) = b_i.
+
+        :return: Gram matrix.
         """
-        pass
+        logger.info("Solving SDP to approximate Gram matrix.")
+
+        n = 2 * self.n_img
+        # Define and solve the CVXPY problem.
+        # Create a symmetric matrix variable.
+        G = cp.Variable((n, n), symmetric=True)
+        # The operator >> denotes matrix inequality.
+        constraints = [G >> 0]
+        constraints += [cp.trace(A[i] @ G) == b[i] for i in range(3 * self.n_img)]
+        prob = cp.Problem(cp.Minimize(cp.trace(-S @ G)), constraints)
+        prob.solve()
 
-    def estimate(self):
+        return G.value
+
+    def _deterministic_rounding(self, Gram):
         """
-        perform estimation of orientations
+        Deterministic rounding procedure to recover the rotations from the Gram matrix.
+
+        The Gram matrix contains information about the first two columns of every rotation
+        matrix. These columns are extracted and used to form the remaining column of every
+        rotation matrix.
+
+        :param Gram: A 2n_img x 2n_img Gram matrix.
+
+        :return: An n_img x 3 x 3 stack of rotation matrices.
         """
-        pass
+        logger.info("Recovering rotations from Gram matrix.")
+
+        # Obtain top eigenvectors from Gram matrix.
+        d, v = stable_eigsh(Gram, 5)
+        sort_idx = np.argsort(-d)
+        logger.info(f"Top 5 eigenvalues from (rank-3) Gram matrix: {d[sort_idx]}")
 
-    def output(self):
+        # Only need the top 3 eigen-vectors.
+        v = v[:, sort_idx[:3]]
+
+        # According to the structure of the Gram matrix, the first `n_img` rows, denoted v1,
+        # correspond to the linear combination of the vectors R_{i}^{1}, i=1,...,K, that is of
+        # column 1 of all rotation matrices. Similarly, the second `n_img` rows of v,
+        # denoted v2, are linear combinations of R_{i}^{2}, i=1,...,K, that is, the second
+        # column of all rotation matrices.
+        v1 = v[: self.n_img].T
+        v2 = v[self.n_img : 2 * self.n_img].T
+
+        # Use a least-squares method to get A.T*A and a Cholesky decomposition to find A.
+        A = self._ATA_solver(v1, v2)
+
+        # Recover the rotations. The first two columns of all rotation
+        # matrices are given by unmixing V1 and V2 using A. The third
+        # column is the cross product of the first two.
+        r1 = np.dot(A.T, v1)
+        r2 = np.dot(A.T, v2)
+        r3 = np.cross(r1, r2, axis=0)
+
+        rotations = np.empty((self.n_img, 3, 3), dtype=self.dtype)
+        rotations[:, :, 0] = r1.T
+        rotations[:, :, 1] = r2.T
+        rotations[:, :, 2] = r3.T
+        # Make sure that we got rotations by enforcing R to be
+        # a rotation (in case the error is large)
+        u, _, v = np.linalg.svd(rotations)
```

Comment:
I should have realized this yesterday in the meeting, but this is not actually finding a best rank-1 approximation (as we do in other places). Here we are only enforcing the orthonormality that we expect this set of rotation matrices to have.

We are essentially enforcing that the singular values of `rotations` are all 1, by factoring into (u * s) @ v.T and then doing the vectorized multiplication u @ v.T with the `einsum` (effectively setting `s` to `1`'s).

So I guess we are finding the `best_orthonormal_approximation` or closest set of actual rotation matrices.

Going to just leave this as is for now unless you think it would be useful to have a `best_orthonormal_approximation` tool.

Reply:
Yes this was my intention. May be easier to make an issue and look at where else it would be used before going through with adding another utility function.

Reply:
Sounds good. Thanks.

> Side note, I also noticed while searching that we already have a `aspire.utils.best_rank1_approximation` method.

Yes this is why I referenced it in my original comment.

Reply:
> Yes this is why I referenced it in my original comment.

My apologies, looks like I completely misunderstood your initial suggestion!